### PR TITLE
Protect against manually updated generated dependencies

### PR DESCRIPTION
**Pull request by potiuk**
The "generated/provider_dependencies.json" is regenerated automatically together with pyproject.toml from all the provider.yaml files. The file contains information about dependencies for all providers and it is used to determine a number of actions when we build providers (for example it is used to automatically generated cross-provider dependencies, and determine whether provider is ready to be released)

The "dependencies" from provider_dependencies.json are also reflected in pyproject.toml file in order to determine what should be installed when you install specific editable extra, so pyproject.toml is also generated automatically together with
provider_dependencies.json. All is good when that generation is done automatically, but so far, when you updated provider_dependencies.json manually, the pyproject.toml was not regenerated - it was actually skipped from regeneration.

This PR changes it by storing hash of provider_depenedencies.json (mixed with the has of the script that generates them) in the generated pyproject.toml file. This way, both - provider_dependencies and pyproject.toml wil always be regenerated by the
"update_providers_dependencies.py" pre-commit whenever there is any inconsistency between provider.yaml files, provider_dependencies.json or pyproject.toml.

Related: #37040

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -533,6 +534,7 @@ alibaba = [
   "oss2>=2.14.0",
 ]
 amazon = [
+  "PyAthena>=3.0.10",
```
Comment (potiuk) [inline pyproject.toml:537]: BTW. It turned out that PyAthena dependency was also manuall added here, which is a bad idea, this part of pyproject.toml will be always now regenerated automatically when things will change.

**Comment by potiuk** [comment]
I also added a small thing - in the generated pyproject.toml I've added comment stating where the source of the information is - pointing to the provider.yaml file where the dependencies are coming from. I hope that will avoid cases where people will try to modify this part manually.

**Review by Taragolis — APPROVED** [review]

**Review by hussein-awala — APPROVED** [review]
Wow, nice one!

**Comment by potiuk** [comment]
> Wow, nice one!

Oh. Yeah. I really, really like seeing someone problems/struggling and turning it into a solution to prevent them.

**Comment by potiuk** [comment]
AH, I need to close that one and open one from the `apache` repo - because we have the case that `dev` scripts that are used in `pull_request_workflow` are modified.

**Comment by potiuk** [comment]
All right. I actually found a way, how to avoid having to do that again: https://github.com/apache/airflow/pull/37057

**Pull request closed without merging**